Client/Frontend/Menu/MenuViewController.swift

```diff
@@ -0,0 +1,194 @@
+        self.popoverPresentationController?.backgroundColor = self.popoverBackgroundColor
+    }
+
+}
+extension MenuViewController: MenuItemDelegate {
+    func menuView(menu: MenuView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
+    }
+}
+
+extension MenuViewController: MenuItemDataSource {
+    func numberOfPagesInMenuView(menuView: MenuView) -> Int {
+        let menuItems = menuConfig.menuItems
+        return Int(ceil(Double(menuItems.count) / Double(maxNumberOfItemsPerPage)))
+    }
+
+    func numberOfItemsPerRowInMenuView(menuView: MenuView) -> Int {
+        return menuConfig.numberOfItemsInRow
+    }
+
+    func menuView(menuView: MenuView, numberOfItemsForPage page: Int) -> Int {
+        let menuItems = menuConfig.menuItems
+        let pageStartIndex = page * maxNumberOfItemsPerPage
+        if (pageStartIndex + maxNumberOfItemsPerPage) > menuItems.count {
+            return menuItems.count - pageStartIndex
+        }
+        return maxNumberOfItemsPerPage
+    }
+
+    func menuView(menuView: MenuView, viewForItemAtIndexPath indexPath: NSIndexPath) -> MenuItemView {
+        let menuItemView = menuView.dequeueReusableMenuItemViewForIndexPath(indexPath)
+
+        let menuItems = menuConfig.menuItems
+        let menuItem = menuItems[indexPath.getMenuItemIndex()]
+
+        menuItemView.menuTitleLabel.text = menuItem.title
+        menuItemView.accessibilityLabel = menuItem.title
+        menuItemView.menuTitleLabel.font = menuConfig.menuFont()
+        menuItemView.menuTitleLabel.textColor = menuConfig.menuTintColorForMode(isPrivate: isPrivate)
+        if let icon = menuItem.iconForMode(isPrivate: isPrivate) {
+            menuItemView.menuImageView.image = icon
+        }
+
+        return menuItemView
+    }
+
+    @objc private func didReceiveLongPress(recognizer: UILongPressGestureRecognizer) {
+    }
+}
+
+extension MenuViewController: MenuToolbarDataSource {
+    func numberOfToolbarItemsInMenuView(menuView: MenuView) -> Int {
+        guard let menuToolbarItems = menuConfig.menuToolbarItems else { return 0}
+        return menuToolbarItems.count
+    }
+
+    func menuView(menuView: MenuView, buttonForItemAtIndex index: Int) -> UIBarButtonItem {
+        // this should never happen - if we don't have any toolbar items then we shouldn't get this far
+        guard let menuToolbarItems = menuConfig.menuToolbarItems else {
+            return UIBarButtonItem(title: nil, style: UIBarButtonItemStyle.Plain, target: nil, action: nil)
+        }
+        let item = menuToolbarItems[index]
+        let toolbarItemView = UIBarButtonItem(image: item.icon, style: UIBarButtonItemStyle.Plain, target: nil, action: nil)
+        toolbarItemView.accessibilityLabel = item.title
+        return toolbarItemView
+    }
+}
+
+extension MenuViewController: MenuToolbarItemDelegate {
+    func menuView(menuView: MenuView, didSelectItemAtIndex index: Int) {
+    }
+}
+
+extension NSIndexPath {
+    func getMenuItemIndex() -> Int {
```

Comment:
This is adding an instance method to every instance of `NSIndexPath`, even those not involved in the menu?

This feels… weird, especially if we're considering publishing the menu as a separate component.